packages/web-components/fast-element/src/utilities.ts

```diff
@@ -0,0 +1,183 @@
+import { ArrayObserver } from "./index.debug.js";
+import { Disposable, isFunction } from "./interfaces.js";
+import type { Notifier, Subscriber } from "./observation/notifier.js";
+import { Observable } from "./observation/observable.js";
+
+interface ObjectVisitor<TVisitorData> {
+    visitObject(object: any, data: TVisitorData): void;
+    visitArray(array: any[], data: TVisitorData): void;
+    visitProperty(object: any, key: PropertyKey, value: any, data: TVisitorData): void;
+}
+
+function shouldTraverse(value: any, traversed: WeakSet<any> | Set<any>) {
+    return (
+        value !== null &&
+        value !== void 0 &&
+        typeof value === "object" &&
+        !traversed.has(value)
+    );
+}
+
+function traverseObject<TVisitorData>(
+    object: any,
+    deep: boolean,
+    visitor: ObjectVisitor<TVisitorData>,
+    data: TVisitorData,
+    traversed: WeakSet<any> | Set<any>
+): void {
+    if (!shouldTraverse(object, traversed)) {
+        return;
+    }
+
+    traversed.add(object);
+
+    if (Array.isArray(object)) {
+        visitor.visitArray(object, data);
+
+        for (const item of object) {
+            traverseObject(item, deep, visitor, data, traversed);
+        }
+    } else {
+        visitor.visitObject(object, data);
+
+        for (const key in object) {
+            const value = object[key];
+            visitor.visitProperty(object, key, value, data);
+
+            if (deep) {
+                traverseObject(value, deep, visitor, data, traversed);
+            }
+        }
+    }
+}
+
+const noop = () => void 0;
+const observed = new WeakSet<any>();
+
+const makeObserverVisitor: ObjectVisitor<undefined> = {
+    visitObject: noop,
+    visitArray: noop,
+    visitProperty(object: any, propertyName: string, value: any): void {
+        Reflect.defineProperty(object, propertyName, {
+            enumerable: true,
+            get() {
+                Observable.track(object, propertyName);
+                return value;
+            },
+            set(newValue: any) {
+                if (value !== newValue) {
```

Comment:
Maybe I'm misunderstanding this, but won't this setter notify when the assigned value is different than the *initial* value (the value when the object was made observable) and not when the value is different than the *previous* value, as Observable.defineProperty does?

Reply:
Once the new value doesn't match the initial value, the variable that was holding the initial value will be updated with the new value. This works because the value passed into the `visitProperty` method is "closed over" by the setter function, effectively making it like an object member field that we can update.

Reply:
oops yep, I missed the `value = newValue` part. Cool thanks!